문나혜: 섹션 4의 미션 제출

moonnahye opened the pull request:
## #️⃣연관된 이슈

#29

## 📝학습 내용

### 🍀String 클래스는 참조형
String 클래스를 비교할때는  `equals()`를 사용해야한다. 

### 🍀String 클래스는 불변객체
따라서 생성후에 절대로 문자열을 변경할 수 없다.
또한 불변으로 설계되었기때문에 사이드 이펙트가 발생하지 않는다!
값을 변경하려면 새로운 String 객체를 생성한다.
수많은 메서드들을 제공한다. 
: length() , equals(), substring() 등등

### 🍀StringBuilder : 가변 String
불변인 String클래스의 단점을 해결하기위해 자바에서 제공한다.

### 🍀String 최적화
자바 컴파일러는 문자열을 `+` 더하는 것을 자동으로 처리해준다.
반복문 안에서 문자끼리 더하면 최적화가 어려울수있는데, 그럴 경우 StringBuilder를 사용하는게 좋다.

### 🍀메서드 체이닝
메서드의 호출결과로 자기자신을 호출하면, `.`을 이용해서 계속 연결해서 사용할 수있는데 이것을 메서드 체이닝이라고 한다. 
StringBuilder를 이용하면 메서드 체이닝으로 계속 연산할 수 있다.


### 스크린샷 (선택)

## 💬리뷰 요구사항

시간 복잡도는 고려하지 못했습니다.. .ᐟ 
첫번째 문제는 sort()메서드를 이용해서 비교하는 것을 사용했고,
두번째 문제는 문자를 하나씩 양쪽에서 하나씩 비교해서 다르면 "" 빈문자열을 리턴하는 걸로 했는데 
더 나은 알고리즘이 있다면 알려주시면 좋을것같아요.ᐟ
그리고 메서드 명도 고민해서 써보았어요!

moonnahye committed:
feat. 미션4 문제 1,2 풀이

Darimi2222 commented on nahye/java-mid1/src/section04/Quiz_1.java:
```diff
@@ -0,0 +1,30 @@
+
+    public static boolean isAnagram(String str1, String str2) {
+
+        if (str1.length() != str2.length()) {
```
공백 무시 조건이 있기 때문에 'Lis   ten' 과 'Silent'도 애너그램으로 판정되어야 합니다!

moonnahye replied:
아 생각을못했네요! 수정하겠습니다!

Darimi2222 commented on nahye/java-mid1/src/section04/Quiz_1.java:
```diff
@@ -0,0 +1,30 @@
+        if (str1.length() != str2.length()) {
+            return false;
+        }
+        str1 = str1.toLowerCase();
```
공백 무시에 대한 부분도 신경써서, 한 가지 더 문자열 조작을 해보는 게 좋을 것 같아요!

Darimi2222 reviewed (commented):
문제 2번은 저도 지식이 부족해서, 1번에 대한 코멘트만 우선 남겨봅니다!

SungbinYang commented on nahye/java-mid1/src/section04/Quiz_1.java:
```diff
@@ -0,0 +1,30 @@
+
+        char[] charArray1 = str1.toCharArray();
+        char[] charArray2 = str2.toCharArray();
+        Arrays.sort(charArray1);
+        Arrays.sort(charArray2);
```
Arrays.sort를 사용하여 문자열을 정렬하는 방식은 𝑂(𝑛log𝑛)의 시간 복잡도를 가집니다. 큰 문자열의 경우 비효율적일 수 있습니다.

moonnahye replied:
수정해보겠습니다!

SungbinYang commented on nahye/java-mid1/src/section04/Quiz_1.java:
```diff
@@ -0,0 +1,30 @@
+        System.out.println(result);
+    }
+
+    public static boolean isAnagram(String str1, String str2) {
```
현재 코드는 영어 소문자만 처리할 수 있습니다. 만약 문자열에 특수 문자, 숫자, 또는 다국어 문자가 포함될 경우, 예상치 못한 동작을 할 수 있습니다.

moonnahye replied:
넵 수정해보겠습니다!

SungbinYang commented on nahye/java-mid1/src/section04/Quiz_2.java:
```diff
@@ -0,0 +1,41 @@
+
+        String longest = "";
+
+       for(int i = 0; i < input.length(); i++) {
+           for (int j = i + 1; j < input.length(); j++) {
```
이중 루프와 extractPalindrome 메서드에서  𝑂(𝑛^2)루프와  O(n) 비교를 수행하여 총 시간 복잡도가 𝑂(𝑛^3)입니다.
긴 문자열의 경우 성능상의 문제가 발생하고 요구사항과도 맞지 않기에 다시 생각해보심이 좋을 것 같습니다!

SungbinYang commented on nahye/java-mid1/src/section04/Quiz_2.java:
```diff
@@ -0,0 +1,41 @@
+
+        while (left < right) {
+            if (input.charAt(left) != input.charAt(right)) {
+                return "";
```
extractPalindrome 메서드는 팰린드롬이 아닐 경우 빈 문자열을 반환합니다. 빈 문자열은 팰린드롬 판별에 불필요하며, 메인 로직이 이를 처리하는 데 추가적인 검증이 필요합니다. 차라리 팰린드롬이 아닌 경우에 대해 바로 종료하지 말고, false 또는 길이 값을 반환하도록 수정하면 효율적일 것 같습니다!

SungbinYang commented on nahye/java-mid1/src/section04/Quiz_2.java:
```diff
@@ -0,0 +1,41 @@
+
+       for(int i = 0; i < input.length(); i++) {
+           for (int j = i + 1; j < input.length(); j++) {
+               String result = extractPalindrome(input, i, j);
```
팰린드롬이 중앙 기준으로 확장된다는 성질을 활용하여 중심에서 확장하는 방식으로 변경하면 중복 계산을 줄일 수 있습니다.

SungbinYang reviewed (commented):
리뷰 남겨드렸습니다! 확인바랍니다~

lazyshu commented on nahye/java-mid1/src/section04/Quiz_1.java:
```diff
@@ -0,0 +1,30 @@
+
+        char[] charArray1 = str1.toCharArray();
+        char[] charArray2 = str2.toCharArray();
+        Arrays.sort(charArray1);
```
흐규 추악한 시간복잡도 ㅠ.ㅠ 저도 이것떄문에 고민많이했는데 저는 그냥.. str2를 StringBuilder로 만들어서 포문 돌렸네용 str1의 첫글자부터 비교해서 str2에 str1에 있는 글자가 있으면 str2에서 그 글자를 뺴주고 존재하지않는다고 하면 false return 했어용
띄어쓰기도 중간에것까지 없애는 쉬운 메서드가 이미 만들어져있더쿤용!!

lazyshu replied:
흙 성빈님이 제꺼도 틀렸다고 하시네염.. 제방식 참고하지 마셔용 ㅜ.ㅠ

lazyshu reviewed (approved):
>_<b

moonnahye committed:
fix. 미션 4의 문제1번 로직 수정 및 피드백 반영 by 성빈님, 채원님

- 공백 처리 추가
- 입력값에 대한 검사 추가
- Arrays.equals() 제거 후 알파벳 배열 추가

SungbinYang reviewed (approved):
수정사항 확인했습니다!

moonnahye committed:
fix. 미션 4의 문제 2번 수정 및 피드백 반영 by 성빈님

- extractPalindrome() 메소드 팰린드롬인 경우만 반환하도록 수정 및 단어 중간부터 양옆으로 검사하도록 수정
- findLongestPalindrome() 단어길이가 홀수일때 짝수일때로 수정

SungbinYang reviewed (approved):
수정하신 부분 확인했습니다! 고생 많으셨습니다~

hmkim199 commented on nahye/java-mid1/src/section04/Quiz_1.java:
```diff
@@ -0,0 +1,57 @@
+        if (str1.isEmpty() || str2.isEmpty()) {
+            return false;
+        }
+        if (!str1.matches("[a-zA-Z\\s]+") || !str2.matches("[a-zA-Z\\s]+")) {
```
한글이나 특수문자 등도 처리 가능하면 좋을 것 같습니다!

moonnahye replied:
영어 아니면 false로 해보았습니다!

hmkim199 reviewed (commented):
코멘트 확인 부탁드립니다~! 고생 많으셨습니다

tdronnie commented on nahye/java-mid1/src/section04/Quiz_1.java:
```diff
@@ -0,0 +1,57 @@
+            alphabet[c - 'a']++;
+        }
+
+        char[] charArray2 = str2.toCharArray();
+        for (char c : charArray2) {
+            alphabet[c - 'a']--;
+        }
```
이 반복문 안에서 alphabet 배열이 1이 아니면 바로 false를 반환할 수도 있을 것 같아요!

tdronnie reviewed (approved):
사소한 코멘트 하나 드렸습니다. 이번 미션 수고하셨습니다. 👍

Darimi2222 reviewed (approved):
수정하신 부분 확인했습니다! 수고 많으셨습니다!
